### PR TITLE
feat: adding new token cache

### PR DESCRIPTION
**Pull request by akouznetchik**
- Added SSO token auth caching, can be enabled via web UI. This idea is originally from https://github.com/jenkinsci/crowd2-plugin/pull/34 and https://github.com/jenkinsci/crowd2-plugin/pull/162 but modified to work. By default, the authentication-filter is mapped to /* therefore all requests including for /static where the header contains an SSO token are validated. This can significantly degrade page loading times depending on latency and Crowd server load. In our testing, after clearing the content cache, enabling this feature improved performance by a factor of 10-20x. The token validation cache will remain valid for the duration of the specified cache TTL, as is the case with other caches.

### CODE REVIEW

**Comment by DuMaM** [comment]
I will release new version with fix only. 
Then after 2 weeks I will do the same with this PR

**Review by DuMaM — APPROVED** [review]

**Comment by DuMaM** [comment]
In this weekend i will merge this if everything works as expected.

File: src/main/resources/de/theit/jenkins/crowd/CrowdSecurityRealm/CacheConfiguration/config_pl.properties [hunk 1]
```diff
@@ -25,3 +25,4 @@
 
 crowd.cache.size=Rozmiar pami\u0119ci podr\u0119cznej
 crowd.cache.TTL=Czas \u017cycia wpis\u00f3w w pami\u0119ci podr\u0119cznej
+crowd.useTokenCache=Przechowuj tokeny uwierzytelniania SSO
```
Comment (DuMaM) [inline src/main/resources/de/theit/jenkins/crowd/CrowdSecurityRealm/CacheConfiguration/config_pl.properties:28]:
```suggestion
crowd.useTokenCache=Wykorzystaj pami\u0119\u0107 podr\u0119czn\u0105 do przechowywania token\u00F3w SSO
```

File: src/main/resources/de/theit/jenkins/crowd/CrowdSecurityRealm/CacheConfiguration/config.jelly [hunk 2]
```diff
@@ -31,4 +31,7 @@ THE SOFTWARE.
     <f:entry field="ttl" title="${%crowd.cache.TTL}">
         <f:select default="300"/>
     </f:entry>
+    <f:entry title="${%crowd.useTokenCache}" field="useTokenCache">
+        <f:checkbox />
+    </f:entry>
```
Comment (DuMaM) [inline src/main/resources/de/theit/jenkins/crowd/CrowdSecurityRealm/CacheConfiguration/config.jelly:36]: Please move this entry bellow `use cache form`
It's confusing to have those two check boxes so separated :D 

Furthermore, if this is an SSO only caching its worth mentioning what are other caches in helper msg

File: src/main/resources/de/theit/jenkins/crowd/CrowdSecurityRealm/CacheConfiguration/config.properties [hunk 3]
```diff
@@ -25,3 +25,4 @@
 
 crowd.cache.size=Used cache size
 crowd.cache.TTL=Cache records timeout
+crowd.useTokenCache=Cache SSO authentication tokens
```
Comment (DuMaM) [inline src/main/resources/de/theit/jenkins/crowd/CrowdSecurityRealm/CacheConfiguration/config.properties:28]: This is the wrong place to do this
Here are located only a value located in DIV, which unfolds itself after checking use cache.

**Review by DuMaM — CHANGES_REQUESTED** [review]

**Comment by DuMaM** [comment]
Ok, I checked this PR.
It introduces a bit of chaos in the UI.
User can have questions if I do have a checkbox for caching SSO tokens, what is under enable caching?

Furthermore, the checkbox for `use sso caching` is in the wrong place. Mostly because it uses some parameters enabled with the `use cache` checkbox.
Could you please add a new PR where you add a checkbox for each type of cache?
This way we can

**Comment by DuMaM** [comment]
@akouznetchik Could you make it work? This checkbox is not rendering.

**Comment by akouznetchik** [comment]
> @akouznetchik Could you make it work? This checkbox is not rendering.

Seems to work for me. Try an `mvn clean` first. Also the cache TTL is in seconds. Do you still wanna do checkbox per cache type?

**Comment by DuMaM** [comment]
> Seems to work for me. Try an mvn clean first. 
This is a bit confusing to me. I'm doing a clean build every time with docker image.
Could you try to start this with `./_start.sh`? Could you please tell me what's the results?

> Also the cache TTL is in seconds.
This should be updated in separated PR.

> Do you still wanna do checkbox per cache type?
Yes, but maybe in feature it's not mandatory now. I misunderstand your change because it was not rendering for me.

**Comment by akouznetchik** [comment]
> @akouznetchik ? :)

Hey man, I’m just on vacation and didn’t grab my laptop, so I won’t be able to give docker a go. I was testing using the Jenkins maven hpi plugin, hpi run, because I have my own crowd server to test against and didn’t need the docker image. I will give this a go next weekend.

**Comment by akouznetchik** [comment]
@DuMaM ok so this pull request is fine. The reason you do not see the changes is because our current build script does not override anything that was already deployed via  `/usr/share/jenkins/ref/plugins/`

See: https://github.com/jenkinsci/docker/blob/master/README.md#installing-custom-plugins

We can fix this by copying the plugin HPI file with `.override` extension.

```
+++ w/casc/Dockerfile
@@ -22,4 +22,4 @@ RUN     jenkins-plugin-cli --plugins configuration-as-code \
                                      monitoring \
                                      jaxb
 
-COPY    --chown=jenkins:jenkins target/crowd2.hpi /usr/share/jenkins/ref/plugins/
+COPY    --chown=jenkins:jenkins target/crowd2.hpi /usr/share/jenkins/ref/plugins/crowd2.hpi.override
```

I will add a separate PR for this. Essentially our latest plugin was being copied from the `target` folder to the staging area, but never re-installed as a previous version was already deployed.

You can verify this if you connect to the docker container and do 

```
ls -lah /var/jenkins_home/plugins/crowd2.hpi
ls -lah /usr/share/jenkins/ref/plugins/crowd2.hpi
```
They will differ.

**Comment by akouznetchik** [comment]
The only issue that remains is 

https://github.com/jenkinsci/crowd2-plugin/pull/169/commits/00a07d125de8729e455eb30516c4b1e1d7d9143b#diff-887c7f64cc1db4c3e718785955bae49b1c680abc40356ffb79e6433ba1680ac2R217

Our values for `cacheTTL` are in seconds, but your latest cache mechanism update put it in minutes. I think this might lead to an almost infinite cache TTL. You want a new PR for this?

**Comment by DuMaM** [comment]
Yes please.

**Comment by DuMaM** [comment]
I will come back from my vacation and retest it.

**Comment by DuMaM** [comment]
@akouznetchik https://github.com/jenkinsci/crowd2-plugin/pull/169#issuecomment-1584408533
Can you prepare this pr? 😉 

To be fair im waiting for it to start test it

**Comment by akouznetchik** [comment]
> i would be grateful if you can hop in and help me maintain this plugin

Would be my pleasure.

**Comment by akouznetchik** [comment]
@DuMaM okay so I opened two PRs, test them out, then let's merge them, and rebase this. Then we need to remove the additional null check that I added, as you see you already re-added it to master. Anything else?

**Comment by DuMaM** [comment]
> > i would be grateful if you can hop in and help me maintain this plugin
> 
> Would be my pleasure.

Please place request on https://github.com/jenkins-infra/helpdesk
I will approve your request :)

**Comment by DuMaM** [comment]
> @DuMaM okay so I opened two PRs, test them out, then let's merge them, and rebase this. Then we need to remove the additional null check that I added, as you see you already re-added it to master. Anything else?

Please rebase this now ;)

**Comment by DuMaM** [comment]
Why you closed this PR?

**Comment by akouznetchik** [comment]
> Why you closed this PR?

Mistake, can you re-open?

**Comment by akouznetchik** [comment]
Let me know when you have done your testing. Also there are 'Changes requested' pending, but all seem resolved, please double check.

**Comment by DuMaM** [comment]
I dont know why but i sitll cannot see this.

**Review by DuMaM — APPROVED** [review]

**Comment by DuMaM** [comment]
lgtm 😉 
did you tested the plugin update scenario?

**Comment by DuMaM** [comment]
We need to check if this is a potential breaking change.
This will introduce a new serialization structure which may break configs stored on the file system.

**Comment by DuMaM** [comment]
Tomorrow I will test this out on my prod env and if everything works, then I will perform release

**Comment by DuMaM** [comment]
@akouznetchik could you please take a look on this PR 
https://github.com/jenkinsci/crowd2-plugin/pull/25
?

**Comment by akouznetchik** [comment]
@DuMaM can we release this?

**Comment by DuMaM** [comment]
Yes we can.

**Comment by DuMaM** [comment]
https://github.com/jenkins-infra/helpdesk/issues/3765
cc: @akouznetchik

**Pull request merged**